### PR TITLE
Week 5 Calendar Project: Find Meeting Query

### PR DESCRIPTION
**Pull request by PriyaVarra**
This code in FindMeetingQuery.java finds and returns all available times from a collection of events for a meeting with required and optional attendees that are long enough for the requested time. 

This is achieved by first filtering out events that don't contain any of the attendees, separating the start and end times for each event, sorting them in ascending order. Then, we iterate through the points and increment a counter every time a start point is encountered and decrement it when an end point is encountered. When this counter is 0,  we know that there are no events at that time, and when it becomes one, we can check if the time between the current time and the last time the counter was 0 is long enough for the requested duration and return it if it is.

FindMeetingQueryTest.java contains unit tests that check edge cases and various scenarios with finding available meeting times. The current version passes all of the tests.

TimePoint.java is a utility class for separating and sorting the time ranges.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,112 @@
...
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // Returns true if an event contains at least one required or optional attendee
+    Predicate<Event> containsAllAttendees = e -> {
```
Comment (saileshsimhadri) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:32]: containsAllAtendees doesn't seem to make sense with what you're doing. Wouldn't containsAnyAttendee make more sense?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,112 @@
...
+
+    Collection<Event> requiredAndOptionalEvents = 
+        filterEvents(containsAllAttendees, events);
+    Collection<TimePoint> requiredAndOptionalPoints =
+        createSortedTimePoints(requiredAndOptionalEvents); 
+    Collection<TimeRange> allAvailableTimes = 
+        getAvailableTimes(requiredAndOptionalPoints, request.getDuration());
+
+    // Return result if optional attendees can be accomodated or there is only one type of attendee
+    if (!allAvailableTimes.isEmpty() || requiredAttendees.isEmpty()
```
Comment (saileshsimhadri) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Why also return it if either are empty?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,112 @@
...
+        || optionalAttendees.isEmpty()) {
+        return allAvailableTimes;
+    }
+
+    // Returns true if an event contains at least one required attendee
+    Predicate<Event> containsRequiredAttendees = e -> {   
+      return !Collections.disjoint(e.getAttendees(), requiredAttendees); 
+    };
+    
+    // If optional attendees cannot be accomodated, try again with only required attendees
```
Comment (saileshsimhadri) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: Is there no need to find the range that maximizes the number of optional attendees?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 4]
```diff
@@ -121,6 +127,182 @@ public void everyAttendeeIsConsidered() {
...
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
```
Comment (saileshsimhadri) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:185]: Why isn't 8:30-9 a valid time?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -121,6 +127,182 @@ public void everyAttendeeIsConsidered() {
...
+    MeetingRequest request =
+        new MeetingRequest(Collections.emptyList(), DURATION_1_HOUR);
+
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Collections.emptyList();
+
+    Assert.assertEquals(expected, actual);
```
Comment (saileshsimhadri) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:250]: assertTrue(actual.isEmpty())

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 4 above)
Comment (saileshsimhadri) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:185]: Oh, because the optional one can't make it. So its either all or just required, not both?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes that would! I have changed that in the latest commits.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Since right now we are only interested in times that accommodate all employees or only required employees, if there are no optional employees, then the exact same events would be filtered and the algorithm would be run again on the same events. If there are no required employees, then there would be no events in the newly filtered events so we would wrongly return that all times are available if we continued.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: That is the optional next step. I wanted to make a pull request for the first two required tasks before I moved on to that.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 4 above)
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:185]: Yes.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 5 above)
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Thanks! I have changed that in the latest commits.

**Comment by saileshsimhadri** [comment]
LGTM.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -121,6 +127,181 @@ public void everyAttendeeIsConsidered() {
...
+  }
+
+
+  @Test
+  public void optionalAttendeeNoRoom() {
+    // Have one person, but make it so that there is just enough room at one point in the day to
+    // have the meeting, but not for optional attendee.
+    //
+    // Events  : |--A--|     |----A----|
+    //                 |-B-|
```
Comment (wjiamin) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:259]: Is there a way to annotate optional B in this diagram?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,112 @@
...
+    return events.stream().filter(filterFn).collect(Collectors.toList());
+  }
+
+  /** Separates start and end points for each event time range and sorts collection. */
+  private Collection<TimePoint> createSortedTimePoints(Collection<Event> events) {
+    ArrayList<TimePoint> points = new ArrayList<TimePoint>();
+
+    for (Event event : events) {
+      TimeRange timeRange = event.getWhen();
+      points.add(new TimePoint(timeRange.start(), true));
```
[the remark this answers is not in the document]
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for the suggestion! I have added this in the latest commits.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (PriyaVarra) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Oops I accidentally missed that! I have fixed this in the latest commit
